NitroxModel/DataStructures/ThreadSafeCollection.cs

```diff
@@ -0,0 +1,318 @@
+                collection = newSet;
+            }
+        }
+
+        public bool RemoveAt(int index)
+        {
+            lock (locker)
+            {
+                IList<T> list = collection as IList<T>;
+                if (list != null)
+                {
+                    list.RemoveAt(0);
+                    return true;
+                }
+                ISet<T> set = collection as ISet<T>;
+                if (set != null)
+                {
+                    T elem = set.ElementAtOrDefault(index);
+                    return elem != null && set.Remove(elem);
+                }
+                return false;
+            }
+        }
+
+        public bool Contains(T item)
+        {
+            lock (locker)
+            {
+                return collection.Contains(item);
+            }
+        }
+
+        public T TryGetValue(int index, out bool succeeded)
+        {
+            lock (locker)
+            {
+                succeeded = false;
+                T result = collection.ElementAtOrDefault(index);
+                succeeded = !Equals(result, default(T));
+                return result;
+            }
+        }
+
+        public List<T> ToList()
+        {
+            lock (locker)
+            {
+                return new List<T>(collection);
+            }
+        }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            lock (locker)
+            {
+                return CreateCopy(collection).GetEnumerator();
+            }
+        }
+
+        /// <summary>
+        ///     Clears the collection and adds the given items.
+        /// </summary>
+        /// <param name="items">Items to add onto the empty collection.</param>
+        public void Set(IEnumerable<T> items)
+        {
+            lock (locker)
+            {
+                collection.Clear();
+                foreach (T item in items)
+                {
+                    collection.Add(item);
+                }
+            }
+        }
+
+        public void RemoveAll(Func<T, bool> predicate)
+        {
+            lock (locker)
+            {
+                foreach (T item in collection)
+                {
+                    if (predicate(item))
+                    {
+                        collection.Remove(item);
+                    }
+                }
+            }
+        }
+
+        public IEnumerable<T> Clone()
+        {
+            lock (locker)
+            {
+                return CreateCopy(collection);
+            }
+        }
+
+        public T Find(Func<T, bool> predicate)
+        {
+            lock (locker)
+            {
+                foreach (T item in collection)
+                {
+                    if (predicate(item))
+                    {
+                        return item;
+                    }
+                }
+            }
+            throw new ArgumentNullException(nameof(predicate), "No item in collection matches select predicate.");
```

Comment:
minor: seems a weird exception type if predicate is not null.

Reply:
Yikes, read the code wrong here. Good catch.